core/src/main/java/org/elasticsearch/cluster/metadata/AutoExpandReplicas.java

```diff
@@ -0,0 +1,92 @@
+package org.elasticsearch.cluster.metadata;
+
+import org.elasticsearch.common.Booleans;
+import org.elasticsearch.common.settings.Setting;
+
+/**
+ * This class acts as a functional wrapper around the <tt>index.auto_expand_replicas</tt> setting.
+ * This setting or rather it's value is expanded into a min and max value which requires special handling
+ * based on the number of datanodes in the cluster. This class handels all the parsing and streamlines the access to these values.
+ */
+final class AutoExpandReplicas {
+    // the value we recognize in the "max" position to mean all the nodes
+    private static final String ALL_NODES_VALUE = "all";
+    public static final Setting<AutoExpandReplicas> SETTING = new Setting<>(IndexMetaData.SETTING_AUTO_EXPAND_REPLICAS, "false", (value) -> {
+        final int min;
+        final int max;
+        if (Booleans.parseBoolean(value, true) == false) {
+            return new AutoExpandReplicas(0, 0, false);
+        }
+        final int dash = value.indexOf('-');
+        if (-1 == dash) {
+            throw new IllegalArgumentException("failed to parse [" + IndexMetaData.SETTING_AUTO_EXPAND_REPLICAS + "] from value: [" + value + "] at index " + dash);
+        }
+        final String sMin = value.substring(0, dash);
+        try {
+            min = Integer.parseInt(sMin);
+        } catch (NumberFormatException e) {
+            throw new IllegalArgumentException("failed to parse [" + IndexMetaData.SETTING_AUTO_EXPAND_REPLICAS + "] from value: [" + value + "] at index "  + dash, e);
+        }
+        String sMax = value.substring(dash + 1);
+        if (sMax.equals(ALL_NODES_VALUE)) {
+            max = Integer.MAX_VALUE;
+        } else {
+            try {
+                max = Integer.parseInt(sMax);
+            } catch (NumberFormatException e) {
+                throw new IllegalArgumentException("failed to parse [" + IndexMetaData.SETTING_AUTO_EXPAND_REPLICAS + "] from value: [" + value + "] at index "  + dash, e);
+            }
+        }
+        return new AutoExpandReplicas(min, max, true);
+    }, true, Setting.Scope.INDEX);
+
+    private final int minReplicas;
+    private final int maxReplicas;
+    private final boolean enabled;
+
+    AutoExpandReplicas(int minReplicas, int maxReplicas, boolean enabled) {
+        if (minReplicas > maxReplicas) {
+            throw new IllegalArgumentException("[" + IndexMetaData.SETTING_AUTO_EXPAND_REPLICAS + "] minReplicas must be =< maxReplicas but wasn't " + minReplicas + " > "  + maxReplicas);
+        }
+        this.minReplicas = minReplicas;
+        this.maxReplicas = maxReplicas;
+        this.enabled = enabled;
+    }
+
+    int getMinReplicas() {
+        return minReplicas;
+    }
+
+    int getMaxReplicas(int numDataNodes) {
+        return Math.min(maxReplicas, numDataNodes-1);
+    }
+
+    @Override
+    public String toString() {
+        return enabled == false ? Boolean.toString(enabled) : minReplicas + "-" + maxReplicas;
```

Comment:
I think this'd be more clear: `enabled ? minReplicas + "-" + maxReplicas : "false"`